platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/consensus/AncestorIterator.java

```diff
@@ -0,0 +1,210 @@
+     * root
+     */
+    private Instant timeReachedRoot;
+    /** becomes false when done and hasNext should return false */
+    private boolean hasNext = false;
+    /** the state of the state machine searching from curr */
+    private IteratorState state;
+    /** is curr a self ancestor of the judge? */
+    private boolean selfAncestor;
+
+    /**
+     * Create an iterator that will iterate over all ancestors of a given event
+     * @param mark the instance to use to mark events as visited
+     */
+    public AncestorIterator(@NonNull final EventVisitedMark mark) {
+        this.mark = mark;
+    }
+
+    /**
+     * The set of ancestors of the given event (the root of the search) for which valid is true.
+     * This will not include a valid ancestor that is only reachable through invalid ancestors.
+     *
+     * @param root the root event whose ancestors should be searched
+     * @param predicate do a depth-first search, but backtrack from any event e where
+     *     valid(e)==false
+     */
+    public void search(@NonNull final EventImpl root, @NonNull final Predicate<EventImpl> predicate) {
+        clear();
+
+        // use the next mark, so that currently marked events will not be found until they are
+        // marked again
+        mark.nextMark();
+        curr = root;
+        valid = predicate;
+        timeReachedRoot = root.getTimeCreated(); // ancestors of curr reached creator then
+        hasNext = true;
+        state = IteratorState.TRAVERSING_SELF_PARENT;
+        selfAncestor = true;
+    }
+
+    private void clear() {
+        stackRef.clear();
+        stackState.clear();
+        stackSelfAncestor.clear();
+        stackTime.clear();
+        curr = null;
+        valid = null;
+        timeReachedRoot = null;
+        hasNext = false;
+        state = null;
+        selfAncestor = false;
+    }
+
+    /**
+     * @return the time when the event last returned by the iterator first reached a self-ancestor
+     *     of the root. to phrase is differently, it is the timestamp of the oldest self-ancestor of
+     *     root that is also an ancestor of the last event returned.
+     */
+    public @NonNull Instant getTime() {
+        return timeReachedRoot;
+    }
+
+    /**
+     * Returns {@code true} if the iteration has more elements. (In other words, returns {@code
+     * true} if {@link #hasNext} would return an element rather than throwing an exception.)
+     *
+     * @return {@code true} if the iteration has more elements
+     */
+    @Override
+    public boolean hasNext() {
+        return hasNext;
+    }
+
+    /**
+     * Returns the next element in the iteration.
+     *
+     * @return the next element in the iteration
+     * @throws NoSuchElementException if the iteration has no more elements
+     */
+    @Override
+    public @NonNull EventImpl next() {
+        if (!hasNext) {
+            throw new NoSuchElementException("no more events left to iterate over");
+        }
```

Comment:
I suggest checking `curr == null` and throwing an exception if true. The search must be initiated by calling `search(...)` first.

Reply:
I am reluctant to change this code since its the same as current develop, which means it has been working fine for a long time. Changing it might introduce unnecessary risk.